applications/FluidDynamicsApplication/custom_utilities/estimate_dt_utilities.cpp

```diff
@@ -0,0 +1,165 @@
+        block_for_each(rModelPart.Elements(), [&](Element& rElement){
+            const double element_cfl = EstimateDtUtility::CalculateElementCFL(rElement, projected_h_func, current_dt);
+            rElement.SetValue(CFL_NUMBER, element_cfl);
+        });
+
+        KRATOS_CATCH("")
+    }
+
+    double EstimateDtUtility::CalculateElementCFL(
+        const Element &rElement,
+        const ElementSizeFunctionType& rElementSizeCalculator,
+        const double Dt)
+    {
+        // Calculate the midpoint velocity
+        const auto& r_geometry = rElement.GetGeometry();
+        const unsigned int n_nodes = r_geometry.PointsNumber();
+        array_1d<double,3> element_vel = r_geometry[0].FastGetSolutionStepValue(VELOCITY);
+        for (unsigned int i = 1; i < n_nodes; ++i) {
+            element_vel += r_geometry[i].FastGetSolutionStepValue(VELOCITY);
+        }
+        element_vel /= static_cast<double>(n_nodes);
+
+        // Calculate element CFL
+        const double h_min = rElementSizeCalculator(r_geometry);
+        const double elem_cfl = norm_2(element_vel) * Dt / h_min;
+
+        return elem_cfl;
+    }
+
+    typename EstimateDtUtility::ElementSizeFunctionType EstimateDtUtility::GetMinimumElementSizeFunction(const Geometry<Node<3>>& rGeometry)
+    {
+        ElementSizeFunctionType projected_h_func;
+        const auto geometry_type = rGeometry.GetGeometryType();
+        if (geometry_type == GeometryData::Kratos_Triangle2D3) {
+            projected_h_func = [&](const Geometry<Node<3>>& rGeometry){return ElementSizeCalculator<2,3>::MinimumElementSize(rGeometry);};
+        } else if (geometry_type == GeometryData::Kratos_Quadrilateral2D4) {
+            projected_h_func = [&](const Geometry<Node<3>>& rGeometry){return ElementSizeCalculator<2,4>::MinimumElementSize(rGeometry);};
+        } else if (geometry_type == GeometryData::Kratos_Tetrahedra3D4) {
+            projected_h_func = [&](const Geometry<Node<3>>& rGeometry){return ElementSizeCalculator<3,4>::MinimumElementSize(rGeometry);};
+        } else if (geometry_type == GeometryData::Kratos_Quadrilateral3D8) {
+            projected_h_func = [&](const Geometry<Node<3>>& rGeometry){return ElementSizeCalculator<3,8>::MinimumElementSize(rGeometry);};
+        } else {
+            KRATOS_ERROR << "Non supported geometry type." << std::endl;
+        }
+
+        return projected_h_func;
+    }
+
+    void EstimateDtUtility::save(Serializer& rSerializer) const
+    {
+        rSerializer.save("mCFL", mCFL);
+        rSerializer.save("mDtMax", mDtMax);
+        rSerializer.save("mDtMin", mDtMin);
+        rSerializer.save("mrModelPart", mrModelPart);
+    }
+
+    void EstimateDtUtility::load(Serializer& rSerializer)
+    {
+        rSerializer.load("mCFL", mCFL);
+        rSerializer.load("mDtMax", mDtMax);
+        rSerializer.load("mDtMin", mDtMin);
+        rSerializer.load("mrModelPart", mrModelPart);
+    }
```

Comment:
out of curiosity, why is this needed?

Reply:
Well, in case you serialize the model for the restart you probably need to keep those no?

Reply:
it only makes sense if this utility is stored e.g. in the `ProcessInfo` (since then it is included in the serialization of the ModelPart)
Usually utilities such as this one are not serialized, same for processes, they are reinitinialized
Consider also that our restart files are large already

Reply:
Agree. This will be always reinitialized. Removing them.